### PR TITLE
Doc 5743 tracing

### PR DESCRIPTION
**Pull request by RichardSmedley**

### CODE REVIEW

File: modules/ROOT/pages/client-settings.adoc [hunk 1]
```diff
@@ -461,6 +463,139 @@ These settings apply to the connection level of a specific bucket configuration
...
+
+.Threshold Logging Tracer Properties
+[#threshold-logging-tracer-ref,cols="3,3,1"]
+|===
+| Property Name | Description | Default Value
+
+| `OperationTracingEnabled`
+| Boolean used to determine tracing is enabled.
+Defaults to using the `ThesholdLoggingTracer` if enabled.
+When false a Noop or similar tracing implementation should be used instead.
```
Comment (MikeGoldsmith) [inline modules/ROOT/pages/client-settings.adoc]: "When false a Noop tracing implementation is used instead."

File: modules/ROOT/pages/client-settings.adoc [hunk 2]
```diff
@@ -461,6 +463,139 @@ These settings apply to the connection level of a specific bucket configuration
...
+| Boolean used to instruct the SDK to try and retrieve duration metrics from the server for KV operations.
+| true
+
+| `ThresholdLoggingTracerInterval`
+| The interval between executions that process the collected operation spans.
+Expressed in milliseconds.
+| 10,000 (10 seconds)
+
+| `ThresholdLoggingTracerSampleSize`
+| The maximum number of items to log per service.
```
Comment (MikeGoldsmith) [inline modules/ROOT/pages/client-settings.adoc]: This is the number of sampled traces to log - maybe something like "The maximum number of sampled traces to log per service."

File: modules/ROOT/pages/client-settings.adoc [hunk 3]
```diff
@@ -461,6 +463,139 @@ These settings apply to the connection level of a specific bucket configuration
...
+Expressed in microseconds.
+| 500,000 (500 milliseconds)
+
+| `ThresholdLoggingTracerViewsThreshold`
+| The View query operation threshold.
+Expressed in microseconds.
+| 1,000,000 (1 second)
+
+| `ThresholdLoggingTracerQueryThreshold`
+| The N1QL query operation threshold.
```
Comment (MikeGoldsmith) [inline modules/ROOT/pages/client-settings.adoc]: Standardised name is just 'query' for n1ql so I'd go with "The query operation threshold."

File: modules/ROOT/pages/client-settings.adoc [hunk 4]
```diff
@@ -461,6 +463,139 @@ These settings apply to the connection level of a specific bucket configuration
...
+Expressed in microseconds.
+| 1,000,000 (1 second)
+
+| `ThresholdLoggingTracerSearchThreshold`
+| The FTS query operation threshold.
```
Comment (MikeGoldsmith) [inline modules/ROOT/pages/client-settings.adoc]: "The search operation threshold."

File: modules/ROOT/pages/client-settings.adoc [hunk 5]
```diff
@@ -461,6 +463,139 @@ These settings apply to the connection level of a specific bucket configuration
...
+Expressed in microseconds.
+| 1,000,000 (1 second)
+
+| `ThresholdLoggingTracerAnalyticsThreshold`
+| The Analytics query operation threshold.
```
Comment (MikeGoldsmith) [inline modules/ROOT/pages/client-settings.adoc:516]: Should _Analytics_ capitalised?

File: modules/ROOT/pages/client-settings.adoc [hunk 6]
```diff
@@ -461,6 +463,139 @@ These settings apply to the connection level of a specific bucket configuration
...
+Expressed in microseconds.
+| 1,000,000 (1 second)
+
+| `OrphanedResponseLoggingEnabled`
+| Boolean used to determine if orphaned response logging is enabled.
+| true
+
+| `OrphanedResponseLoggingInterval`
+| The interval used to flush orphaned response information to the log.
```
Comment (MikeGoldsmith) [inline modules/ROOT/pages/client-settings.adoc]: I like the ThresholdLogger description style better, something like:
"The interval between executions that process orphaned responses to the log."

File: modules/ROOT/pages/client-settings.adoc [hunk 7]
```diff
@@ -461,6 +463,139 @@ These settings apply to the connection level of a specific bucket configuration
...
+Expressed in microseconds.
+| 10,000 (10 seconds)
+
+| `OrphanedResponseLoggingSampleSize`
+| The number of sample orphaned responses whose to log additional information for per execution.
```
Comment (MikeGoldsmith) [inline modules/ROOT/pages/client-settings.adoc]: Same as ThresholdLogger description:
"The maximum number of sampled orphaned responses to log."

File: modules/ROOT/pages/client-settings.adoc [hunk 8]
```diff
@@ -461,6 +463,139 @@ These settings apply to the connection level of a specific bucket configuration
...
+| 10
+|===
+
+
+
+Applying these as client-level settings with App.Config (or Web.Config) can be done as follows:
```
Comment (MikeGoldsmith) [inline modules/ROOT/pages/client-settings.adoc]: Don't need capitalisation, should just be web.config / web.config.

File: modules/ROOT/pages/tracing-from-the-sdk.adoc [hunk 9]
```diff
@@ -156,23 +156,33 @@ This tells us the following:
...
 Useful to figure out which node is affected.
-* *last_local_id:* With Server 5.5 and later, this id is negotiated with the server and can be used to correlate logging information on both sides in a simpler fashion.
-* *last_dispatch_us:* The time when the client sent the request and got the response took around 1 millisecond.
-* *total_us:* The total time it took to perform the full operation: here around 1.5 milliseconds.
-* *dispatch_us:* The amount of time observed between sending the request over the network to when a response was received.
-* *server_us:* The server reported that its work performed took 22 microseconds (this does not include network time or time in the buffer before picked up at the cluster).
-* *decode_us:* Decoding the response took the client 12 microseconds.
+* *last_local_id:* The last connection ID used to send a packet to the server.
+With Server 5.5 and later, this id is negotiated with the server and can be used to correlate logging information on both sides in a simpler fashion.
+* *last_dispatch_us:* The time taken for the last dispatch to the server: around 1 millisecond.
```
Comment (MikeGoldsmith) [inline modules/ROOT/pages/tracing-from-the-sdk.adoc:165]: As an operation may be retried, it's important to indicate that *last*_ properties are for the most recent socket hostname & port used to send the operation.

File: modules/ROOT/pages/tracing-from-the-sdk.adoc [hunk 10]
```diff
@@ -156,23 +156,33 @@ This tells us the following:
...
+* *server_us:* The calculated sum of all server duration sub-spans -- only present if server durations are enabled.
+Here the server reported that its work performed took 22 microseconds (this does not include network time or time in the buffer before picked up at the cluster).
+* *decode_us:* The calculated sum of all decode sub-spans.
+Here, decoding the response took the client 12 microseconds.
 
 You can see that if the thresholds are set the right way based on production requirements, without much effort slow operations can be logged and pinpointed more easily than before.
 
+NOTE: Fields like *total_us* give a time that includes not just the operation, but all possible retries (which the SDK will attempt in response to errors caused by node failure or rebalancing), so will often be longer than a single operation such as that given by *last_dispatch_us*.
+
+// TODO look into the Java bug whereby one of them is not aggregated
```
Comment (MikeGoldsmith) [inline modules/ROOT/pages/tracing-from-the-sdk.adoc]: Should remove TODO for java follow up 😄

**Review by MikeGoldsmith — CHANGES_REQUESTED** [review]

**Review by MikeGoldsmith — APPROVED** [review]

File: modules/ROOT/pages/client-settings.adoc (hunk 5 above)
Comment (RichardSmedley) [reply modules/ROOT/pages/client-settings.adoc:516]: Yes, as with Query and Search, here referring directly to one of our services.

File: modules/ROOT/pages/tracing-from-the-sdk.adoc (hunk 9 above)
Comment (RichardSmedley) [reply modules/ROOT/pages/tracing-from-the-sdk.adoc:165]: Thanks - added this as note to `last_local_address`

**Pull request merged**